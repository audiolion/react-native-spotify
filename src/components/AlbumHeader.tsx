import React from 'react';
import { Colors } from '../helpers/constants';
import {
  Dimensions,
  StyleSheet,
  ImageBackground,
  SafeAreaView,
  View,
  Text,
} from 'react-native';
import { material } from 'react-native-typography';

type Props = {
  uri: string;
};

const oneQuarterScreenHeight = Math.floor(Dimensions.get('screen').height / 4);

export const AlbumHeader: React.FC<Props> = ({ uri }) => {
  const source = { uri };
  return (
    <ImageBackground
      source={source}
      style={styles.background}
      imageStyle={styles.image}
    >
      <SafeAreaView style={styles.headerContainer}>
        <View style={styles.center}>
          <Text style={[material.headlineWhite, styles.headline]}>
            Saved Albums
          </Text>
        </View>
      </SafeAreaView>
    </ImageBackground>
  );
};

const styles = StyleSheet.create({
  background: {
    height: oneQuarterScreenHeight,
    width: Dimensions.get('window').width,
    shadowRadius: 12,
    shadowOpacity: 0.13,
    shadowColor: Colors.tabBarShadowColor,
    shadowOffset: { height: 6, width: 0 },
  },
  image: {
    opacity: 0.7,
  },
  headerContainer: {
    flexDirection: 'row',
    flex: 1,
    height: oneQuarterScreenHeight,
  },
  center: {
    position: 'absolute',
    opacity: 1,
    ...StyleSheet.absoluteFillObject,
    justifyContent: 'center',
    alignItems: 'center',
  },
  headline: {
    fontSize: 30,
    fontWeight: 'bold',
    backgroundColor: 'rgba(0,0,0,0.7)',
    paddingVertical: 10,
    paddingHorizontal: 20,
    borderRadius: 20,
  },
});
