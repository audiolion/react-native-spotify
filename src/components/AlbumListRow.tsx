import React from 'react';
import {
  View,
  TouchableHighlight,
  Text,
  StyleSheet,
  Image,
} from 'react-native';
import Icon from 'react-native-vector-icons/Entypo';
import { material } from 'react-native-typography';
import { Colors } from '../helpers/constants';

type Props = {
  item: Echobind.SavedAlbum;
  onPress: () => void;
};

export const AlbumListRow: React.FC<Props> = ({ item, onPress }) => {
  return (
    <TouchableHighlight onPress={onPress}>
      <View style={styles.listRow}>
        <Image
          source={{ uri: item.album.images[0].url }}
          style={styles.image}
        />
        <View style={styles.flexStart}>
          <Text style={material.title} numberOfLines={1} ellipsizeMode="tail">
            {item.album.name}
          </Text>
        </View>
        <Icon name="chevron-thin-right" color={Colors.spotifyGreen} size={20} />
      </View>
    </TouchableHighlight>
  );
};

const styles = StyleSheet.create({
  listRow: {
    paddingHorizontal: 20,
    paddingVertical: 10,
    backgroundColor: Colors.spotifyWhite,
    borderBottomColor: Colors.spotifyBlack,
    borderBottomWidth: StyleSheet.hairlineWidth,
    flexDirection: 'row',
    flex: 1,
    alignItems: 'center',
    justifyContent: 'space-between',
  },
  image: {
    height: 40,
    width: 40,
    marginRight: 10,
  },
  flexStart: {
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'flex-start',
  },
});
