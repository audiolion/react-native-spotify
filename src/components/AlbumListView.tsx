import React from 'react';
import { StyleSheet, FlatList, Text, View } from 'react-native';
import {
  NavigationScreenProp,
  NavigationRoute,
  NavigationParams,
} from 'react-navigation';
import { material } from 'react-native-typography';
import { TouchableHighlight } from 'react-native-gesture-handler';

type Props = {
  navigation: NavigationScreenProp<
    NavigationRoute<NavigationParams>,
    NavigationParams
  >;
  name: string;
  albums: any[];
  onRefresh: () => void;
  refreshing: boolean;
};

export const AlbumListView: React.FC<Props> = ({
  navigation,
  name,
  albums,
  onRefresh,
  refreshing,
}) => {
  const keyExtractor = (item: any) =>
    (item && item.album && item.album.id) || `${Math.random() * -1}`;

  const goToAlbumNotes = (id: string | null) => () =>
    id && id.charAt(0) !== '-' && navigation.navigate('AlbumNotes');

  return (
    <View style={styles.container}>
      <Text style={material.title}>{name}'s Albums</Text>
      <FlatList
        data={albums}
        onRefresh={onRefresh}
        refreshing={refreshing}
        keyExtractor={keyExtractor}
        renderItem={({ item }) =>
          item &&
          item.album && (
            <View>
              <TouchableHighlight onPress={goToAlbumNotes(item.album.id)}>
                <Text style={material.body1}>{item.album.name}</Text>
              </TouchableHighlight>
            </View>
          )
        }
      />
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
});
