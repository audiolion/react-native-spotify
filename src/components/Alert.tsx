import React from 'react';
import { TextStyle, ViewProps, Text, StyleSheet, View } from 'react-native';
import { Colors } from '../helpers/constants';

export interface AlertProps extends ViewProps {
  message: string;
  textStyle?: TextStyle;
  type?: 'red' | 'yellow' | 'blue';
}

export const Alert: React.FC<AlertProps> = props => {
  const type = props.type || 'blue';
  const alertTextStyle =
    props.type === 'yellow' ? styles.alertTextDark : styles.alertTextLight;
  return (
    <View
      // @ts-ignore
      style={[styles.alertContainer, styles[`${type}Container`], props.style]}
    >
      <Text style={[alertTextStyle, props.textStyle]}>{props.message}</Text>
      {props.children}
    </View>
  );
};

const styles = StyleSheet.create({
  alertContainer: {
    marginVertical: 10,
    marginHorizontal: 16,
    borderRadius: 4,
    paddingVertical: 10,
    paddingHorizontal: 16,
  },
  alertTextLight: {
    color: Colors.spotifyWhite,
  },
  alertTextDark: {
    color: Colors.spotifyBlack,
  },
  redContainer: {
    backgroundColor: Colors.red,
  },
  yellowContainer: {
    backgroundColor: Colors.yellow,
  },
  blueContainer: {
    backgroundColor: Colors.blue,
  },
});
