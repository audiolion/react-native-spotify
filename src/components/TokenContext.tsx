import React from 'react';

export const TokenContext = React.createContext<
  [string | null, React.Dispatch<React.SetStateAction<string | null>>]
>([null, () => {}]);

export const TokenProvider: React.FC = ({ children }) => {
  const state = React.useState<string | null>(null);

  return (
    <TokenContext.Provider value={state}>{children}</TokenContext.Provider>
  );
};
