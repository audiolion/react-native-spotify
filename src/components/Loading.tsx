import React from 'react';
import { useTimeout } from '@ryanar/hooks';
import {
  ActivityIndicator,
  View,
  StyleSheet,
  StyleProp,
  ViewStyle,
} from 'react-native';
import { Alert } from './Alert';
import { Colors } from '../helpers/constants';

type Props = {
  delay?: number;
  error?: boolean;
  timedOut?: boolean;
  style?: StyleProp<ViewStyle>;
  animating?: boolean;
  color?: string;
  size?: number | 'small' | 'large';
};

export const Loading: React.FC<Props> = ({
  delay = 1000,
  error = undefined,
  timedOut = false,
  style,
  animating = true,
  color,
  size,
}) => {
  const shouldDisplay = useTimeout(delay);

  if (error) {
    return (
      <Alert
        message="Data could not be retrieved."
        type="red"
        style={styles.row}
      />
    );
  }

  if (timedOut) {
    return (
      <Alert
        message="Loading is taking a long time."
        type="yellow"
        style={styles.row}
      />
    );
  }

  if (!shouldDisplay) {
    return null;
  }

  return (
    <View style={[styles.container, style]}>
      <ActivityIndicator
        animating={animating}
        testID={
          process.env.NODE_ENV === 'test'
            ? 'components-loading-indicator'
            : undefined
        }
        color={color || Colors.spotifyGreen}
        size={size}
      />
    </View>
  );
};

const styles = StyleSheet.create({
  container: { alignItems: 'center', justifyContent: 'center' },
  row: {
    flexDirection: 'row',
    alignItems: 'center',
  },
  retryButton: {
    backgroundColor: 'transparent',
    paddingTop: 2,
    paddingBottom: 2,
    paddingRight: 2,
    paddingLeft: 8,
    borderWidth: 0,
  },
});
