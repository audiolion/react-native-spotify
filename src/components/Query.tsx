import React from 'react';
import { DocumentNode } from 'graphql';
import { NavigationScreenProp } from 'react-navigation';
import { OperationVariables } from '@apollo/react-common';
import { useQuery as useApolloQuery } from '@apollo/react-hooks';
import { QueryComponentOptions } from '@apollo/react-components';
import { Loading } from './Loading';
import { NetworkStatus } from 'apollo-client';
import AsyncStorage from '@react-native-community/async-storage';

interface Props {
  query: DocumentNode;
  navigation: NavigationScreenProp<any>;
  errorFallback?: React.ReactElement | null;
  onLoading?: React.ReactElement | null;
  retry?: () => void;
}

export function Query<TData = any, TVariables = OperationVariables>({
  query,
  children,
  navigation,
  onLoading,
  errorFallback,
  retry,
  ...options
}: Props & Omit<QueryComponentOptions<TData, TVariables>, 'query'>) {
  const result = useApolloQuery<TData, TVariables>(query, {
    fetchPolicy: 'cache-and-network',
    ...options,
  });

  if (
    result.loading &&
    ![NetworkStatus.fetchMore, NetworkStatus.refetch].includes(
      result.networkStatus
    )
  ) {
    return onLoading || <Loading />;
  }

  if (result.error) {
    const { graphQLErrors } = result.error;
    const isUnauthenticated = graphQLErrors.find(
      error => error.message && error.message === 'Unauthorized'
    );
    if (isUnauthenticated) {
      AsyncStorage.multiRemove(['access-token', 'token-expires']).then(() =>
        navigation.navigate('Auth')
      );
      return null;
    }
    return errorFallback || <Loading error />;
  }

  return children(result);
}
