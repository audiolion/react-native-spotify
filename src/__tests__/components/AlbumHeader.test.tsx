import React from 'react';
import { render } from 'react-native-testing-library';
import { AlbumHeader } from '../../components/AlbumHeader';

describe('AlbumHeader', () => {
  it('matches snapshot', () => {
    const tree = render(
      <AlbumHeader title="foo" uri="https://example.com" />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
