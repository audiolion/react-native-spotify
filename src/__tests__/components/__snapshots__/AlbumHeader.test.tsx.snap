// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AlbumHeader matches snapshot 1`] = `
<View
  accessibilityIgnoresInvertColors={true}
  style={
    Object {
      "height": 333,
      "shadowColor": "#32325d",
      "shadowOffset": Object {
        "height": 6,
        "width": 0,
      },
      "shadowOpacity": 0.13,
      "shadowRadius": 12,
      "width": 750,
    }
  }
>
  <Image
    source={
      Object {
        "uri": "https://example.com",
      }
    }
    style={
      Array [
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        Object {
          "height": 333,
          "width": 750,
        },
        Object {
          "opacity": 0.7,
        },
      ]
    }
  />
  <RCTSafeAreaView
    emulateUnlessSupported={true}
    style={
      Object {
        "flex": 1,
        "flexDirection": "row",
        "height": 333,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "bottom": 0,
          "justifyContent": "center",
          "left": 0,
          "opacity": 1,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "backgroundColor": "transparent",
              "color": "rgba(255,255,255,1)",
              "fontFamily": "System",
              "fontSize": 24,
              "fontWeight": "400",
              "letterSpacing": 0.351562,
              "lineHeight": 32,
            },
            Object {
              "backgroundColor": "rgba(0,0,0,0.7)",
              "borderRadius": 20,
              "fontSize": 30,
              "fontWeight": "bold",
              "paddingHorizontal": 20,
              "paddingVertical": 10,
            },
          ]
        }
      >
        foo
      </Text>
    </View>
  </RCTSafeAreaView>
</View>
`;
